beacon-chain/p2p/gossipsub_peer_crawler.go

```diff
@@ -0,0 +1,515 @@
+	// we will discover it during a re-crawl and ping it again to test for reachability.
+	// we're not blacklisting this peer anyways.
+	cp.updateTopicsUnlocked(existingPNode, nil)
+}
+
+func (cp *crawledPeers) updateCrawledIfNewer(node *enode.Node, topics []string) {
+	cp.mu.Lock()
+
+	enodeID := node.ID()
+	existingPNode, ok := cp.byEnode[enodeID]
+
+	if ok && existingPNode.node == nil {
+		log.WithField("enodeId", enodeID).Error("enode is nil for enodeId")
+		cp.mu.Unlock()
+		return
+	}
+
+	// we don't want to update enodes with a lower sequence number as they're stale records
+	if ok && existingPNode.node.Seq() >= node.Seq() {
+		cp.mu.Unlock()
+		return
+	}
+
+	if !ok {
+		// this is a new peer
+		peerID, err := enodeToPeerID(node)
+		if err != nil {
+			log.WithError(err).WithField("node", node.ID()).Debug("Failed to convert enode to peer ID")
+			cp.mu.Unlock()
+			return
+		}
+		existingPNode = &peerNode{
+			id:     enodeID,
+			node:   node,
+			peerID: peerID,
+			topics: make(map[gossipsubcrawler.Topic]struct{}),
+		}
+		cp.byEnode[enodeID] = existingPNode
+		cp.byPeerId[peerID] = existingPNode
+	} else {
+		existingPNode.node = node
+	}
+
+	cp.updateTopicsUnlocked(existingPNode, topics)
+
+	if existingPNode.isPinged || len(topics) == 0 {
+		cp.mu.Unlock()
+		return
+	}
+	cp.mu.Unlock()
+
+	select {
+	case cp.g.pingCh <- *node:
+	case <-cp.g.ctx.Done():
+		return
+	}
+}
+
+func (cp *crawledPeers) removeTopic(topic gossipsubcrawler.Topic) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+
+	// Get all peers subscribed to this topic
+	peers, ok := cp.byTopic[topic]
+	if !ok {
+		return // Topic doesn't exist
+	}
+
+	// Remove the topic from each peer's topic list
+	for peerID := range peers {
+		if pnode, exists := cp.byPeerId[peerID]; exists {
+			delete(pnode.topics, topic)
+			// remove the peer if it has no more topics left
+			if len(pnode.topics) == 0 {
+				cp.updateTopicsUnlocked(pnode, nil)
+			}
+		}
+	}
+
+	// Remove the topic from byTopic map
+	delete(cp.byTopic, topic)
+}
+
+func (cp *crawledPeers) removePeerId(peerID peer.ID) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+
+	pnode, exists := cp.byPeerId[peerID]
+	if !exists {
+		return
+	}
+
+	// Use updateTopicsUnlocked with empty topics to remove the peer
+	cp.updateTopicsUnlocked(pnode, nil)
+}
+
+func (cp *crawledPeers) removePeer(enodeID enode.ID) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+	pnode, exists := cp.byEnode[enodeID]
+	if !exists {
+		return
+	}
+	cp.updateTopicsUnlocked(pnode, nil)
+}
+
+// setting topics to empty will remove the peer completely.
+func (cp *crawledPeers) updateTopicsUnlocked(pnode *peerNode, topics []string) {
+	// If topics is empty, remove the peer completely.
+	if len(topics) == 0 {
+		delete(cp.byPeerId, pnode.peerID)
+		delete(cp.byEnode, pnode.id)
+		for t := range pnode.topics {
+			if peers, ok := cp.byTopic[t]; ok {
+				delete(peers, pnode.peerID)
+				if len(peers) == 0 {
+					delete(cp.byTopic, t)
+				}
+			}
+		}
```

Comment:
Nit: I have a slight preference for the version with less code in deep scope
But for sure both versions are fine.


```suggestion
		for t := range pnode.topics {
			peers, ok := cp.byTopic[t]
			if !ok {
				continue
			}

			delete(peers, pnode.peerID)
			if len(peers) == 0 {
				delete(cp.byTopic, t)
			}
		}
```

Reply:
Can we keep the original version ? I don't have a strong preference either way.